examples/multigpu_advtrain/resnet_tf.py

```diff
@@ -0,0 +1,354 @@
+        for i in six.moves.range(1, self.hps.num_residual_units):
+            with tf.variable_scope('unit_3_%d' % i):
+                x = res_func(x, filters[3], filters[3],
+                             self._stride_arr(1), False)
+
+        with tf.variable_scope('unit_last'):
+            x = self._batch_norm('final_bn', x)
+            x = self._relu(x, self.hps.relu_leakiness)
+            x = self._global_avg_pool(x)
+
+        with tf.variable_scope('logit'):
+            logits = self._fully_connected(x, self.hps.num_classes)
+            self.predictions = tf.nn.softmax(logits)
+
+        return logits, self.predictions
+
+    def create_sync_ops(self, host_device):
+        sync_ops = []
+        for layer in self.layers:
+            sync_ops += layer.create_sync_ops(host_device)
+        return sync_ops
+
+    def build_cost(self, labels, logits):
+        op = logits.op
+        if "softmax" in str(op).lower():
+            logits, = op.inputs
+
+        with tf.variable_scope('costs'):
+            xent = tf.nn.softmax_cross_entropy_with_logits(
+                logits=logits, labels=labels)
+            cost = tf.reduce_mean(xent, name='xent')
+            cost += self._decay()
+            cost = cost
+
+        return cost
+
+    def build_train_op_from_cost(self, cost):
+        """Build training specific ops for the graph."""
+        self.lrn_rate = tf.constant(self.hps.lrn_rate, tf.float32,
+                                    name='learning_rate')
+        self.momentum = tf.constant(self.hps.momentum, tf.float32,
+                                    name='momentum')
+
+        trainable_variables = tf.trainable_variables()
+        grads = tf.gradients(cost, trainable_variables)
+
+        if self.hps.optimizer == 'sgd':
+            optimizer = tf.train.GradientDescentOptimizer(self.lrn_rate)
+        elif self.hps.optimizer == 'adam':
+            optimizer = tf.train.AdamOptimizer(learning_rate=self.lrn_rate)
+        elif self.hps.optimizer == 'mom':
+            optimizer = tf.train.MomentumOptimizer(self.lrn_rate,
+                                                   self.momentum)
+
+        # there should be no gradients wrt vars on other gpus
+        gv_pairs = zip(grads, trainable_variables)
+        gv_pairs = [gv for gv in gv_pairs if gv[0] is not None]
+        devs = set([gv[1].device for gv in gv_pairs])
+        assert len(devs) == 1
+
+        apply_op = optimizer.apply_gradients(
+            gv_pairs,
+            global_step=self.global_step, name='train_step')
+
+        train_ops = [apply_op] + self._extra_train_ops
+        train_op = tf.group(*train_ops)
+        return train_op
+
+    def _build_train_op(self, logits, labels):
+        self.cost = self.build_cost(labels, logits)
+
+        train_op = self.build_train_op_from_cost(self.cost)
+
+        return train_op
+
+    # TODO(xpan): Consider batch_norm in contrib/layers/python/layers/layers.py
```

Comment:
it probably doesn't make sense to preserve TODOs when duplicating open source code

Reply:
Fixed in 45cdaf7.